pkg/driver/microphone/microphone_windows.go

```diff
@@ -0,0 +1,329 @@
+package microphone
+
+import (
+	"errors"
+	"golang.org/x/sys/windows"
+	"io"
+	"time"
+	"unsafe"
+
+	"github.com/pion/mediadevices/pkg/driver"
+	"github.com/pion/mediadevices/pkg/io/audio"
+	"github.com/pion/mediadevices/pkg/prop"
+)
+
+// Windows APIs
+var (
+	winmm                 = windows.NewLazySystemDLL("Winmm.dll")
+	waveInOpen            = winmm.NewProc("waveInOpen")
+	waveInStart           = winmm.NewProc("waveInStart")
+	waveInStop            = winmm.NewProc("waveInStop")
+	waveInReset           = winmm.NewProc("waveInReset")
+	waveInClose           = winmm.NewProc("waveInClose")
+	waveInPrepareHeader   = winmm.NewProc("waveInPrepareHeader")
+	waveInAddBuffer       = winmm.NewProc("waveInAddBuffer")
+	waveInUnprepareHeader = winmm.NewProc("waveInUnprepareHeader")
+)
+
+type buffer struct {
+	waveHdr
+	data []int16
+}
+
+func newBuffer(samples int) *buffer {
+	b := make([]int16, samples)
+	return &buffer{
+		waveHdr: waveHdr{
+			lpData:         uintptr(unsafe.Pointer(&b[0])),
+			dwBufferLength: uint32(samples * 2),
+		},
+		data: b,
+	}
+}
+
+type microphone struct {
+	hWaveIn windows.Pointer
+	buf     map[uintptr]*buffer
+	chBuf   chan *buffer
+	closed  chan struct{}
+}
+
+func init() {
+	// TODO: enum devices
+	driver.GetManager().Register(&microphone{}, driver.Info{
+		Label:      "default",
+		DeviceType: driver.Microphone,
+	})
+}
+
+func (m *microphone) Open() error {
+	m.chBuf = make(chan *buffer)
+	m.buf = make(map[uintptr]*buffer)
+	m.closed = make(chan struct{})
+	return nil
+}
+
+func (m *microphone) cbWaveIn(hWaveIn windows.Pointer, uMsg uint, dwInstance, dwParam1, dwParam2 *int32) uintptr {
+	switch uMsg {
+	case MM_WIM_DATA:
+		b := m.buf[uintptr(unsafe.Pointer(dwParam1))]
+		m.chBuf <- b
+
+	case MM_WIM_OPEN:
+	case MM_WIM_CLOSE:
+		close(m.chBuf)
+	}
+	return 0
+}
+
+func (m *microphone) Close() error {
+	if m.hWaveIn == nil {
+		return nil
+	}
+	close(m.closed)
+
+	ret, _, _ := waveInStop.Call(
+		uintptr(unsafe.Pointer(m.hWaveIn)),
+	)
+	if err := errWinmm[ret]; err != nil {
+		return err
+	}
+	ret, _, _ = waveInReset.Call(
+		uintptr(unsafe.Pointer(m.hWaveIn)),
+	)
+	if err := errWinmm[ret]; err != nil {
+		return err
+	}
+	for _, buf := range m.buf {
+		ret, _, _ := waveInUnprepareHeader.Call(
+			uintptr(unsafe.Pointer(m.hWaveIn)),
+			uintptr(unsafe.Pointer(&buf.waveHdr)),
+			uintptr(unsafe.Sizeof(buf.waveHdr)),
+		)
+		if err := errWinmm[ret]; err != nil {
+			return err
+		}
+	}
+	ret, _, _ = waveInClose.Call(
+		uintptr(unsafe.Pointer(m.hWaveIn)),
+	)
+	if err := errWinmm[ret]; err != nil {
+		return err
+	}
+	<-m.chBuf
+	m.hWaveIn = nil
+
+	return nil
+}
+
+func (m *microphone) AudioRecord(p prop.Media) (audio.Reader, error) {
+	for i := 0; i < 5; i++ {
+		b := newBuffer(
+			int(uint64(p.Latency) * uint64(p.SampleRate) / uint64(time.Second)),
+		)
+		m.buf[uintptr(unsafe.Pointer(&b.waveHdr))] = b
+	}
```

Comment:
I'm not sure what this for loop is being used for. Why are we creating the same buffer 5 times?

Reply:
This API needs multiple buffering (at least double buffer) to get signal continuously.

Reply:
hmm, but, why 5 I guess?

Reply:
It preserves received buffer until `audio.Reader.Read` is called.
5 has no specific reason, but it means that the maximum blockable duration to get data without dropping chunks is 5 * `p.Latency` = 100ms at now. `Read` timing jitter or delay larger than this will make the sound choppy.

Reply:
Ah, I see, do you mind to add this as a comment?

Reply:
I'll make it named constant and add a comment.

Reply:
thanks!

Reply:
Added some more comments especially about buffering.